### PR TITLE
Add support for running integration tests agains docker-compose

### PR DESCRIPTION
**Pull request by bodymindarts**
This PR introduces a new way to run the test suite and is intended as a first step in improving the overall developer workflow.

Problems with existing execution method:
- minikube setup was time-consuming and error prone
- a lot of unnecessary complexity / cognitive overhead due to having the entire k8s stack as part of the test-execution runtime

Key concepts introduced in this PR:
- Add a `Makefile` for grouping top level commands. Currently `yarn` is used as a task executor but that has many drawbacks.
- Add a `.envrc` file for loading required environment via [Direnv](https://direnv.net)
- Add a `docker-compose` file for bringing up runtime deps in a lightweight fashion.
- Start writing dev documentation

Some things to note:
- For now this is introduced as a completely separate and alternative workflow. I don't want to risk breaking the existing way of doing things because I don't know where all it is relied upon. Once there are no more dependencies on the existing workflow we should just delete it all at once.
- A slight modification to the production code was required to enable this in the `src/redis.ts` file.

### CODE REVIEW

**Comment by nicolasburtey** [comment]
> Add a Makefile for grouping top level commands. Currently yarn is used as a task executor but that has many drawbacks.

what drawback do you see? I'd rather avoiding using several task executor. and I've not seen much if any javascript/typescript project using Makefile AFAIR

**Comment by nicolasburtey** [comment]
there is only 1 tls/macaroon but we have 3 in our tests. have not run the test yet but I think this will be an issue

**Comment by bodymindarts** [comment]
> what drawback do you see? I'd rather avoiding using several task executor. and I've not seen much if any javascript/typescript project using Makefile AFAIR

I think `yarn` makes sense as a task driver for stuff that interfaces with the JS/TS world. But for generic tasks that are not JS specific (eg. executing deploys, or running integration tests) I prefer to use a JS agnostic task driver.

This has quite a few benefits. Here are some I can think of right away:
- Editors have formatting and syntax highlighting vs having to put code into a `json` that is just displayed as a JSON string.
- Because the way `Makefiles` are used (as language agnostic task runners) developers from other eco-systems that find the repo will be quickly find an entry point from which to see what top-level tasks to execute (rather than having to find them in the language specific `package.json`
- Tasks that don't require a JS runtime / `yarn` installation (ie. deploying to an environment, or running end-to-end tests) can be run without needing `nodjs` installed. In particular from within CI
- Uniformity across repos in the organisation. ie. when we have other repos for spinning up / opperating infrastructure (for example) we would not be using `yarn` as a task driver.... if we have a `Makefile` in every repo there is less cognitive overhead for context switching between repos.

> there is only 1 tls/macaroon but we have 3 in our tests. have not run the test yet but I think this will be an issue
I am re-using the `lnd` config across all `lnd` service in the `docker-compose` file:
```  lnd:
    image: lncm/lnd:v0.12.1-monitoring
    ports:
      - "10009:10009"
    volumes:
    - ${PWD}/dev/lnd/lnd.conf:/data/.lnd/lnd.conf
    - ${PWD}/dev/lnd/tls.key:/data/.lnd/tls.key
    - ${PWD}/dev/lnd/tls.cert:/data/.lnd/tls.cert
    depends_on: [bitcoind]
  lnd-outside-1:
    image: lncm/lnd:v0.12.1-monitoring
    ports:
      - "10010:10009"
    volumes:
    - ${PWD}/dev/lnd/lnd.conf:/data/.lnd/lnd.conf
    - ${PWD}/dev/lnd/tls.key:/data/.lnd/tls.key
    - ${PWD}/dev/lnd/tls.cert:/data/.lnd/tls.cert
    depends_on: [bitcoind]
  lnd-outside-2:
    image: lncm/lnd:v0.12.1-monitoring
    ports:
      - "10011:10009"
    volumes:
    - ${PWD}/dev/lnd/lnd.conf:/data/.lnd/lnd.conf
    - ${PWD}/dev/lnd/tls.key:/data/.lnd/tls.key
    - ${PWD}/dev/lnd/tls.cert:/data/.lnd/tls.cert
    depends_on: [bitcoind]
 ```
 
 I don't see any benefits of having unique `tls` keys and `lnd.conf` files for each other than keeping more code in the repository.

File: DEV.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+$ git clone git@github.com:GaloyMoney/galoy.git
+$ cd galoy
+$ yarn install
+```
+
+We use [direnv](https://direnv.net) to load environment variables needed for running the integration tests.
+```
+$ direnv allow
+direnv reload
+direnv: direnv: loading ~/projects/GaloyMoney/galoy/.envrc
```
Comment (nicolasburtey) [inline DEV.md:15]: not clear if this a command that need to be type or an output here

Comment (nicolasburtey) [reply DEV.md:15]: I guess an output but `direnv allow` return nothing on my end

File: DEV.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+(...)
+```
+
+To execute the test suite runtime dependencies must be running.
+```
+$ make start-deps
+docker-compose up -d
```
Comment (nicolasburtey) [inline DEV.md:22]: missing $

File: DEV.md (hunk 1 above)
Comment (bodymindarts) [reply DEV.md:15]: its output. Lines starting with `$ ...` are input lines.

File: DEV.md (hunk 2 above)
Comment (bodymindarts) [reply DEV.md:22]: No its output

Comment (nicolasburtey) [reply DEV.md:22]: ok

Comment (bodymindarts) [reply DEV.md:22]: Make by default echos the commands its running back to the shell

File: DEV.md (hunk 1 above)
Comment (bodymindarts) [reply DEV.md:15]: > I guess an output but direnv allow return nothing on my end

Thats strange... I have:
```
$ direnv --version
2.28.0
```

File: DEV.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+```
+$ direnv allow
+direnv reload
+direnv: direnv: loading ~/projects/GaloyMoney/galoy/.envrc
+(...)
+```
+
+To execute the test suite runtime dependencies must be running.
+```
+$ make start-deps
```
Comment (nicolasburtey) [inline DEV.md:21]: I have this at this stage:

```
$ make start-deps
docker-compose up -d
Traceback (most recent call last):
  File "urllib3/connectionpool.py", line 670, in urlopen
  File "urllib3/connectionpool.py", line 392, in _make_request
  File "http/client.py", line 1255, in request
  File "http/client.py", line 1301, in _send_request
  File "http/client.py", line 1250, in endheaders
  File "http/client.py", line 1010, in _send_output
  File "http/client.py", line 950, in send
  File "docker/transport/unixconn.py", line 43, in connect
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "requests/adapters.py", line 439, in send
  File "urllib3/connectionpool.py", line 726, in urlopen
  File "urllib3/util/retry.py", line 410, in increment
  File "urllib3/packages/six.py", line 734, in reraise
  File "urllib3/connectionpool.py", line 670, in urlopen
  File "urllib3/connectionpool.py", line 392, in _make_request
  File "http/client.py", line 1255, in request
  File "http/client.py", line 1301, in _send_request
  File "http/client.py", line 1250, in endheaders
  File "http/client.py", line 1010, in _send_output
  File "http/client.py", line 950, in send
  File "docker/transport/unixconn.py", line 43, in connect
urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "docker/api/client.py", line 214, in _retrieve_server_version
  File "docker/api/daemon.py", line 181, in version
  File "docker/utils/decorators.py", line 46, in inner
  File "docker/api/client.py", line 237, in _get
  File "requests/sessions.py", line 543, in get
  File "requests/sessions.py", line 530, in request
  File "requests/sessions.py", line 643, in send
  File "requests/adapters.py", line 498, in send
requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "docker-compose", line 3, in <module>
  File "compose/cli/main.py", line 81, in main
  File "compose/cli/main.py", line 199, in perform_command
  File "compose/cli/command.py", line 60, in project_from_options
  File "compose/cli/command.py", line 152, in get_project
  File "compose/cli/docker_client.py", line 41, in get_client
  File "compose/cli/docker_client.py", line 170, in docker_client
  File "docker/api/client.py", line 197, in __init__
  File "docker/api/client.py", line 221, in _retrieve_server_version
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
[15104] Failed to execute script docker-compose
make: *** [start-deps] Error 255
```

Comment (bodymindarts) [reply DEV.md:21]: Which docker-compose version are you running:
```
$ docker-compose --version
Docker Compose version 2.0.0-beta.3
```

Maybe its outdated. I'm using syntax `version: "3"` in the file.

Comment (bodymindarts) [reply DEV.md:21]: Which docker-compose version are you running:
```
$ docker-compose --version
Docker Compose version 2.0.0-beta.3
```

Maybe its outdated. I'm using syntax `version: "3"` in the file.

Comment (nicolasburtey) [reply DEV.md:21]: `$ docker-compose --version
docker-compose version 1.29.0, build 07737305`

File: DEV.md (hunk 1 above)
Comment (nicolasburtey) [reply DEV.md:15]: same version

File: DEV.md (hunk 3 above)
Comment (bodymindarts) [reply DEV.md:21]: Made a commit fixing that. I wonder if that helps with your old `docker-compose` version

Comment (nicolasburtey) [reply DEV.md:21]: same output

Comment (bodymindarts) [reply DEV.md:21]: Very strange... hmm.
Perhaps it doesn't like the env interpolation:

```
    volumes:
    - ${PWD}/dev/lnd/lnd.conf:/data/.lnd/lnd.conf
```

Comment (bodymindarts) [reply DEV.md:21]: Pushed another commit

**Comment by nicolasburtey** [comment]
if that helps

```
$ direnv status
direnv exec path /usr/local/bin/direnv
DIRENV_CONFIG /Users/nb/.config/direnv
bash_path /usr/local/bin/bash
disable_stdin false
warn_timeout 5s
whitelist.prefix []
whitelist.exact map[]
No .envrc loaded
Found RC path /Users/nb/Code/galoy/galoy-app2/galoy/.envrc
Found watch: ".envrc" - 2021-06-18T13:45:09+01:00
Found watch: "../../../../.local/share/direnv/allow/d413e3e1c71777468a483f5d5618388da33d8490e3fdc3e7bc84d9f9dbb6d7cf" - 2021-06-18T13:27:20+01:00
Found RC allowed true
Found RC allowPath /Users/nb/.local/share/direnv/allow/d413e3e1c71777468a483f5d5618388da33d8490e3fdc3e7bc84d9f9dbb6d7cf
```

**Comment by bodymindarts** [comment]
Pushed some commits with the things we discussed @nicolasburtey.

Would be great if @krtk6160 could verify that its also working on his setup.

**Pull request merged**